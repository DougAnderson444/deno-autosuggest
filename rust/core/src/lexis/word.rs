use std::fmt;
use std::iter::empty;
pub use crate::lang::Lang;
pub use super::{Chars, CharPattern};
pub use super::pos::PartOfSpeech;


#[derive(PartialEq)]
pub struct Word<T: AsRef<[char]>> {
    pub source: T,
    pub slice:  (usize, usize),
    pub chars:  T,
    pub stem:   usize,
    pub pos:    Option<PartOfSpeech>,
    pub fin:    bool,
}


impl<T: AsRef<[char]>> Word<T> {
    pub fn is_primary(&self) -> bool {
        match self.pos {
            Some(PartOfSpeech::Article)     => false,
            Some(PartOfSpeech::Preposition) => false,
            Some(PartOfSpeech::Conjunction) => false,
            Some(PartOfSpeech::Particle)    => false,
            _ => true,
        }
    }
}


impl Word<Vec<char>> {
    pub fn from_vec(source: Vec<char>) -> Word<Vec<char>> {
        let len = source.len();
        let chars = source.clone();
        Word {
            source,
            slice: (0, len),
            chars,
            stem: len,
            pos: None,
            fin: true,
         }
    }

    pub fn from_str(source: &str) -> Word<Vec<char>> {
        let source = source.chars().collect::<Vec<_>>();
        let chars  = source.clone();
        let len    = source.len();
        Word {
            source,
            slice: (0, len),
            chars,
            stem: len,
            pos:  None,
            fin:  true,
         }
    }

    pub fn to_ref<'a>(&'a self) -> Word<&'a [char]> {
        Word {
            source: &self.source,
            slice:  self.slice,
            chars:  &self.chars,
            stem:   self.stem,
            pos:    self.pos,
            fin:    self.fin,
        }
    }

    pub fn split<'a, 'b, P: CharPattern>(&'a self, pattern: &'b P) -> WordSplit<'a, 'b, P> {
        WordSplit { word: self, offset: 0, pattern }
    }

    pub fn strip<P: CharPattern>(&mut self, pattern: &P) -> &mut Self {
        let left = self.chars.iter()
            .take_while(|&&ch| pattern.matches(ch))
            .count();
        let right = self.chars.iter()
            .rev()
            .take_while(|&&ch| pattern.matches(ch))
            .take(self.chars.len() - left)
            .count();

        self.chars.splice(self.chars.len() - right .., empty());
        self.chars.splice(0 .. left, empty());

        self.slice = (self.slice.0 + left, self.slice.1 - right);
        self.fin = self.fin || right != 0;
        self
    }

    pub fn stem(&mut self, lang: &Lang) -> &mut Self {
        self.stem = lang.stem(&self.chars);
        self
    }

    pub fn mark_pos(&mut self, lang: &Lang) -> &mut Self {
        self.pos = lang.get_pos(&self.chars);
        self
    }

    pub fn lower(&mut self) -> &mut Self {
        if self.chars.iter().any(|ch| ch.is_uppercase()) {
            for ch in &mut self.chars {
                *ch = ch.to_lowercase().next().unwrap();
            }
        }
        self
    }
}


impl<'a> Word<&'a [char]> {
    pub fn to_own(&self) -> Word<Vec<char>> {
        Word {
            source: self.source.to_vec(),
            slice:  self.slice,
            chars:  self.chars.to_vec(),
            stem:   self.stem,
            pos:    None,
            fin:    self.fin,
        }
    }
}


impl<T: AsRef<[char]>> Word<T> {
    pub fn fin(mut self, fin: bool) -> Self {
        self.fin = fin;
        self
    }

    pub fn len(&self) -> usize {
        self.chars.as_ref().len()
    }

    pub fn is_empty(&self) -> bool {
        self.chars.as_ref().is_empty()
    }
}


impl<T: AsRef<[char]>> fmt::Debug for Word<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Word {{ \"")?;
        let len  = self.chars.as_ref().len();
        let stem = self.stem;
        for (i, ch) in self.chars.as_ref().iter().enumerate() {
            write!(f, "{}", ch)?;
            if i == stem - 1 && i != len - 1 {
                write!(f, "|")?;
            }
        }
        write!(f, "\"")?;
        if !self.fin { write!(f, "..")?; }
        write!(f, " }}")?;
        Ok(())
    }
}


#[derive(Debug)]
pub struct WordSplit<'b, 'c, P: CharPattern> {
    word: &'b Word<Vec<char>>,
    offset: usize,
    pattern: &'c P,
}


impl<'b, 'c, P: CharPattern> Iterator for WordSplit<'b, 'c, P> {
    type Item = Word<&'b [char]>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.offset >= self.word.len() {
            return None;
        }

        self.offset += self.word.chars[self.offset ..]
            .iter()
            .take_while(|&&ch| self.pattern.matches(ch))
            .count();

        let len = self.word.chars[self.offset ..]
            .iter()
            .take_while(|&&ch| !self.pattern.matches(ch))
            .count();

        if len == 0 {
            return None;
        }

        let word = Word {
            source: &self.word.chars[self.offset .. self.offset + len],
            chars:  &self.word.source[self.offset .. self.offset + len],
            slice:  (0, len),
            stem:   len,
            pos:    None,
            fin:    self.word.fin || self.offset + len < self.word.len(),
        };

        self.offset += word.len();

        Some(word)
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::lang::lang_english;
    use super::{Word, Chars, PartOfSpeech};

    use Chars::{
        Whitespaces,
        Punctuation,
    };

    #[test]
    fn word_split() {
        let w = Word::from_str(" Foo Bar, Baz; ");
        let s = w.split(&[Whitespaces, Punctuation]).collect::<Vec<_>>();
        assert_debug_snapshot!(s);
    }

    #[test]
    fn word_split_empty() {
        let w = Word::from_str(" ,;");
        let s = w.split(&[Whitespaces, Punctuation]).collect::<Vec<_>>();
        assert_debug_snapshot!(s);
    }

    #[test]
    fn word_split_unfinished() {
        let mut w1 = Word::from_str(" Foo Bar, Baz");
        let mut w2 = Word::from_str(" Foo Bar, Baz; ");
        w1.fin = false;
        w2.fin = false;
        let s1 = w1.split(&[Whitespaces, Punctuation]).collect::<Vec<_>>();
        let s2 = w2.split(&[Whitespaces, Punctuation]).collect::<Vec<_>>();
        assert_eq!(s1.last().unwrap().fin, false);
        assert_eq!(s2.last().unwrap().fin, true);
    }

    #[test]
    fn word_strip() {
        let mut w = Word::from_str(" Foo Bar, Baz; ");
        w.strip(&[Whitespaces, Punctuation]);
        assert_debug_snapshot!(&w);
        assert_debug_snapshot!(&w.slice);
    }

    #[test]
    fn word_strip_empty() {
        let mut w = Word::from_str(" ,;");
        w.strip(&[Whitespaces, Punctuation]);
        assert_debug_snapshot!(w);
    }

    #[test]
    fn word_strip_unfinished() {
        let mut w1 = Word::from_str(" Foo Bar, Baz");
        let mut w2 = Word::from_str(" Foo Bar, Baz; ");
        w1.fin = false;
        w2.fin = false;
        w1.strip(&[Whitespaces, Punctuation]);
        w2.strip(&[Whitespaces, Punctuation]);
        assert_eq!(w1.fin, false);
        assert_eq!(w2.fin, true);
    }

    #[test]
    fn word_stem() {
        let lang = lang_english();
        let mut w = Word::from_str("university");
        w.stem(&lang);
        assert_eq!(w.stem, 7);
    }

    #[test]
    fn word_pos() {
        let lang = lang_english();
        let mut w1 = Word::from_str("university");
        let mut w2 = Word::from_str("the");
        w1.mark_pos(&lang);
        w2.mark_pos(&lang);
        assert_eq!(w1.pos, None);
        assert_eq!(w2.pos, Some(PartOfSpeech::Article));
    }

    #[test]
    fn word_lower() {
        let mut w = Word::from_str(" Foo Bar, Baz; ");
        w.lower();
        assert_debug_snapshot!(w);
    }
}
