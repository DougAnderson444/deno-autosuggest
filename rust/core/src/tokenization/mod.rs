mod word;
mod text;

use crate::lang::{Lang, Chars, CharPattern};
pub use word::Word;
pub use text::Text;


pub fn tokenize_query(source: &str, lang: &Option<Lang>) -> Text<Vec<char>> {
    Text::from_str(source)
        .normalize(lang)
        .fin(false)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
        .mark_pos(lang)
        .stem(lang)
}


pub fn tokenize_record(source: &str, lang: &Option<Lang>) -> Text<Vec<char>> {
     Text::from_str(source)
        .normalize(lang)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
        .mark_pos(lang)
        .stem(lang)
}
