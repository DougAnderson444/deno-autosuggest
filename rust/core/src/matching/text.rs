use std::cmp::Ordering::{Equal, Less};
use std::cell::RefCell;
use fnv::{FnvHashMap as HashMap};
use crate::tokenization::{Word, TextRef};
use super::WordMatch;
use super::word::word_match;

thread_local! {
    static RMATCHES: RefCell<HashMap<usize, WordMatch>> = RefCell::new(HashMap::with_capacity_and_hasher(10, Default::default()));
    static QMATCHES: RefCell<HashMap<usize, WordMatch>> = RefCell::new(HashMap::with_capacity_and_hasher(10, Default::default()));
}


pub fn text_match(rtext: &TextRef, qtext: &TextRef) -> (Vec<WordMatch>, Vec<WordMatch>) {
    RMATCHES.with(|rcell| {
    QMATCHES.with(|qcell| {
        let rmatches = &mut *rcell.borrow_mut();
        let qmatches = &mut *qcell.borrow_mut();
        rmatches.clear();
        qmatches.clear();

        for qword in qtext.words.iter() {
            if qmatches.contains_key(&qword.offset) { continue; }
            let qword = qword.to_view(qtext);

            let mut candidate: Option<(WordMatch, WordMatch)> = None;

            for rword in rtext.words.iter() {
                if rmatches.contains_key(&rword.offset) { continue; }
                let rword = rword.to_view(rtext);
                let mut stop = false;

                None.or_else(|| {
                        let rnext = rtext.words.get(rword.offset + 1)?.to_view(rtext);
                        if qword.len() < rword.len() + rword.dist(&rnext) { return None; }
                        if rmatches.contains_key(&(rword.offset + 1)) { return None; }
                        let (rmatch,  qmatch)  = word_match(&rword.join(&rnext), &qword)?;
                        let (rmatch1, rmatch2) = rmatch.split(&rword, &rnext)?;
                        rmatches.insert(rmatch1.offset, rmatch1);
                        rmatches.insert(rmatch2.offset, rmatch2);
                        qmatches.insert(qmatch.offset,  qmatch);
                        candidate.take();
                        stop = true;
                        Some(())
                    })
                    .or_else(|| {
                        let qnext = qtext.words.get(qword.offset + 1)?.to_view(qtext);
                        if rword.len() < qword.len() + qword.dist(&qnext) { return None; }
                        if qmatches.contains_key(&(qword.offset + 1)) { return None; }
                        let (rmatch,  qmatch)  = word_match(&rword, &qword.join(&qnext))?;
                        let (qmatch1, qmatch2) = qmatch.split(&qword, &qnext)?;
                        rmatches.insert(rmatch.offset,  rmatch);
                        qmatches.insert(qmatch1.offset, qmatch1);
                        qmatches.insert(qmatch2.offset, qmatch2);
                        candidate.take();
                        stop = true;
                        Some(())
                    })
                    .or_else(|| {
                        let (rmatch2, qmatch2) = word_match(&rword, &qword)?;
                        let score2 = rmatch2.match_len() - 2 * (rmatch2.typos.ceil() as usize);
                        let score1 = candidate
                            .as_ref()
                            .map(|(m, _)| m.match_len() - 2 * (m.typos.ceil() as usize))
                            .unwrap_or(0);
                        let replace = match (candidate.as_ref(), score1.cmp(&score2)) {
                            (None, _) => true,
                            (Some(_), Less) => true,
                            (Some(_), Equal) if !rmatch2.func => true,
                            _ => false,
                        };
                        if replace {
                            stop      = !rmatch2.func;
                            candidate = Some((rmatch2, qmatch2));
                        }
                        Some(())
                    });

                if stop {
                    break;
                }
            }

            if let Some((rmatch, qmatch)) = candidate {
                rmatches.insert(rmatch.offset, rmatch);
                qmatches.insert(qmatch.offset, qmatch);
            }
        }

        let mut rmatches = rmatches.drain().map(|(_, m)| m).collect::<Vec<_>>();
        let mut qmatches = qmatches.drain().map(|(_, m)| m).collect::<Vec<_>>();
        rmatches.sort_by(|m1, m2| m1.offset.cmp(&m2.offset));
        qmatches.sort_by(|m1, m2| m1.offset.cmp(&m2.offset));

        (rmatches, qmatches)
    }) })
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::tokenization::{Text, TextOwn};
    use crate::lang::{CharClass, lang_basic, lang_english, lang_spanish};
    use super::{text_match};


    fn text(s: &str) -> TextOwn {
        let lang = lang_basic();
        Text::from_str(s)
            .split(&[CharClass::Punctuation, CharClass::Whitespace], &lang)
            .set_char_classes(&lang)
    }


    #[test]
    fn match_text_empty_both() {
        let rtext = text("");
        let qtext = text("").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_empty_one() {
        let rtext = text("");
        let qtext = text("mailbox").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
        assert_debug_snapshot!(text_match(&qtext.to_ref(), &rtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_equality() {
        let rtext = text("mailbox");
        let qtext = text("mailbox").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_typos() {
        let rtext = text("mailbox");
        let qtext = text("maiblox").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_first() {
        let rtext = text("yellow mailbox");
        let qtext = text("yelow").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_second() {
        let rtext = text("yellow mailbox");
        let qtext = text("maiblox").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_unfinished() {
        let rtext = text("yellow mailbox");
        let qtext = text("maiblox yel").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_intersection() {
        let rtext = text("small yellow metal mailbox");
        let qtext = text("big malibox yelo").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_best_rword_first() {
        let rtext = text("theory theme");
        let qtext = text("the").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_best_rword_nonfunc() {
        let lang  = lang_english();
        let rtext = text("the theme").set_pos(&lang);
        let qtext = text("the").fin(false).set_pos(&lang);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_best_rword_typos() {
        let lang  = lang_spanish();
        let rtext = text("Cepillo de dientes").set_pos(&lang);
        let qtext = text("de").fin(false).set_pos(&lang);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_regression_best_match() {
        let rtext = text("sneaky");
        let qtext = text("sneak").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_query() {
        let rtext = text("wifi router");
        let qtext = text("wi fi router").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_query_unfihished() {
        let rtext = text("microbiology");
        let qtext = text("micro bio").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_query_typos() {
        let rtext = text("microbiology");
        let qtext = text("mcro byology").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_query_short() {
        let rtext = text("t-light");
        let qtext = text("tli").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_record() {
        let rtext = text("wi fi router");
        let qtext = text("wifi router").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_record_typos() {
        let rtext = text("micro biology");
        let qtext = text("mcrobiology").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_record_unfinished() {
        let rtext = text("micro biology");
        let qtext = text("microbio").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_regression_1() {
        let rtext = text("special, year");
        let qtext = text("especiall").fin(false);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }

    #[test]
    fn match_text_joined_regression_2() {
        let rtext1 = text("50's");
        let rtext2 = text("500w");
        let qtext  = text("50s").fin(false);
        assert_debug_snapshot!(text_match(&rtext1.to_ref(), &qtext.to_ref()));
        assert_debug_snapshot!(text_match(&rtext2.to_ref(), &qtext.to_ref()));
    }
}
