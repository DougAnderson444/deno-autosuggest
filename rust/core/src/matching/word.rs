use crate::tokenization::Word;
use super::WordMatch;
use super::damlev::DamerauLevenshtein;
use super::jaccard::Jaccard;


const LENGTH_THRESHOLD:  f64 = 0.26;
const JACCARD_THRESHOLD: f64 = 0.41;
const DAMLEV_THRESHOLD:  f64 = 0.26;

thread_local! {
    static DAMLEV:  DamerauLevenshtein = DamerauLevenshtein::new();
    static JACCARD: Jaccard<char>      = Jaccard::new();
}


pub fn word_match(rword: &Word, qword: &Word, rchars: &[char], qchars: &[char]) -> Option<WordMatch> {
    if qword.is_empty() || rword.is_empty() {
        return None;
    }
    if !length_check(rword, qword) {
        return None;
    }
    if !jaccard_check(rword, qword, rchars, qchars) {
        return None;
    }

    let rchars = rword.view(rchars);
    let qchars = qword.view(qchars);
    let mut best_match: Option<WordMatch> = None;

    DAMLEV.with(|damlev| {
        damlev.distance(qchars, rchars);
        let dists = &*damlev.dists.borrow();

        let left  = if qword.fin { max!(qword.stem, rword.stem) } else { qword.stem } - 1;
        let right = max!(qword.len(), rword.len()) + 1;

        if right <= left { return; }

        let range = (left .. right).rev();

        for rlen in range.clone() {
            for qlen in range.clone() {
                // Out of bounds.
                if qlen > qword.len() { continue; }
                if rlen > rword.len() { continue; }
                // Left margin is for insertion/deletion, not for both prefixes at the same time.
                if rlen == left && qlen == left  { continue; }
                // Compare full words only if query is finished.
                if qword.fin && rlen < rword.stem { break; }
                if qword.fin && qlen < qword.stem { break; }
                // Words with 2+ insertions/deletions are mismatched by default.
                if (qlen as isize - rlen as isize).abs() > 1 { continue; }

                let dist = dists.get(qlen + 1, rlen + 1);
                let rel  = dist as f64 / max!(qlen, rlen, 1) as f64;

                if rel > DAMLEV_THRESHOLD { continue; }

                best_match = best_match
                    .take()
                    .filter(|m| m.typos <= dist)
                    .or_else(|| Some(WordMatch::new(
                        qword,
                        rword,
                        qlen,
                        rlen,
                        dist,
                    )));

                if dist == 0 {
                    break;
                }
            }
        }
    });

    best_match
}


pub fn length_check(rword: &Word, qword: &Word) -> bool {
    let qlen  = qword.len();
    let rlen  = if qword.fin { rword.len() } else { min!(qlen, rword.len()) };

    if qlen <= 1 || rlen <= 1 {
        return qlen == rlen;
    }

    let long  = max!(qlen, rlen);
    let short = min!(qlen, rlen);
    let dist  = 1.0 - (short as f64 / long as f64);

    dist < LENGTH_THRESHOLD
}


pub fn jaccard_check(rword: &Word, qword: &Word, rchars: &[char], qchars: &[char]) -> bool {
    let rchars = rword.view(rchars);
    let qchars = qword.view(qchars);
    let rslice = if qword.fin { &rchars } else { &rchars[.. min!(qword.len(), rword.len())] };
    let dist   = JACCARD.with(|j| j.rel_dist(&rslice, &qchars));
    dist < JACCARD_THRESHOLD
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::utils::to_vec;
    use crate::tokenization::{Chars, Word, Text};
    use crate::lang::lang_english;
    use super::{length_check, jaccard_check, word_match};


    fn text(s: &str) -> Text<Vec<char>> {
        Text::from_str(s).split(&Chars::Whitespaces)
    }

    #[test]
    fn length_check_len_4() {
        let sample = [
            (false, "m"),
            (false, "ma"),
            (true,  "mai"),
            (true,  "mail"),
            (true,  "mailb"),
            (false, "mailbo"),
            (false, "mailbox"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query);
            let result = length_check(&rtext.words[0], &qtext.words[0]);
            assert_eq!(result, expect, "Failed length_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn length_check_len_7() {
        let sample = [
            (false, "m"),
            (false, "ma"),
            (false, "mai"),
            (false, "mail"),
            (false, "mailb"),
            (true,  "mailbo"),
            (true,  "mailbox"),
            (true,  "mailboxe"),
            (true,  "mailboxes"),
            (false, "mailboxess"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mailbox");
            let qtext  = text(query);
            let result = length_check(&rtext.words[0], &qtext.words[0]);
            assert_eq!(result, expect, "Failed length_check(\"mailbox\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_len_4() {
        let sample = [
            (true,  "mail"),
            (true,  "bail"),
            (false, "bait"),
            (false, "balt"),
            (false, "bolt"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query);
            let result = jaccard_check(&rtext.words[0], &qtext.words[0], &rtext.chars, &qtext.chars);
            assert_eq!(result, expect, "Failed jaccard_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_len_7() {
        let sample = [
            (true,  "mailbox"),
            (true,  "mailbot"),
            (false, "railbot"),
            (false, "raidbot"),
            (false, "roidbot"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mailbox");
            let qtext  = text(query);
            let result = jaccard_check(&rtext.words[0], &qtext.words[0], &rtext.chars, &qtext.chars);
            assert_eq!(result, expect, "Failed jaccard_check(\"mailbox\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_len_7_reduction() {
        let sample = [
            (true,  "mailbox"),
            (true,  "mailbxx"),
            (true,  "mailxxx"),
            (false, "maixxxx"),
            (false, "maxxxxx"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mailbox");
            let qtext  = text(query);
            let result = jaccard_check(&rtext.words[0], &qtext.words[0], &rtext.chars, &qtext.chars);
            assert_eq!(result, expect, "Failed jaccard_check(\"mailbox\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_unfinished() {
        let sample = [
            (true,  "m"),
            (true,  "ma"),
            (true,  "mai"),
            (true,  "mail"),
            (true,  "mailb"),
            (true,  "mailbo"),
            (false, "mailbox"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query).fin(false);
            let result = jaccard_check(&rtext.words[0], &qtext.words[0], &rtext.chars, &qtext.chars);
            assert_eq!(result, expect, "Failed jaccard_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn length_check_unfinished() {
        let sample = [
            (true,  "m"),
            (true,  "ma"),
            (true,  "mai"),
            (true,  "mail"),
            (true,  "mailb"),
            (false, "mailbo"),
            (false, "mailbox"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query).fin(false);
            let result = length_check(&rtext.words[0], &qtext.words[0]);
            assert_eq!(result, expect, "Failed length_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    // Match word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_empty_both() {
        let qchars = to_vec("");
        let rchars = to_vec("");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    #[test]
    fn match_word_empty_record() {
        let qchars = to_vec("mailbox");
        let rchars = to_vec("");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    #[test]
    fn match_word_empty_query() {
        let qchars = to_vec("");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    // Match word: prefixed by same length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_full_strict() {
        let qchars = to_vec("mailbox");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_insertion() {
        let qchars = to_vec("mailybox");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_deletion() {
        let qchars = to_vec("mailox");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_transposition() {
        let qchars = to_vec("maiblox");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_query_too_long() {
        let qchars1 = to_vec("mailboxes");
        let qchars2 = to_vec("mailboxes");
        let rchars  = to_vec("mail");
        let qword1  = Word::new(qchars1.len()).fin(true);
        let qword2  = Word::new(qchars2.len()).fin(false);
        let rword   = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword1, &rchars[..], &qchars1[..]));
        assert_debug_snapshot!(word_match(&rword, &qword2, &rchars[..], &qchars2[..]));
    }

    #[test]
    fn match_word_full_stem() {
        let     rchars  = to_vec("universe");
        let     qchars1 = to_vec("university");
        let     qchars2 = to_vec("university");
        let mut rword   = Word::new(rchars.len());
        let mut qword1  = Word::new(qchars1.len());
        let     qword2  = Word::new(qchars2.len());

        let lang = lang_english();
        qword1.stem(&qchars1[..], &lang);
        rword .stem(&rchars[..],  &lang);

        assert_debug_snapshot!(word_match(&rword, &qword1, &rchars[..], &qchars1[..]));
        assert_debug_snapshot!(word_match(&rword, &qword2, &rchars[..], &qchars2[..]));
    }


    // Match word: prefixed by lesser length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_partial_strict() {
        let qchars = to_vec("mailb");
        let rchars = to_vec("mailbox");
        let qword = Word::new(qchars.len()).fin(false);
        let rword = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_insertion() {
        let qchars = to_vec("maiylb");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_deletion() {
        let qchars = to_vec("maib");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_transposition() {
        let qchars = to_vec("malib");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }
}
