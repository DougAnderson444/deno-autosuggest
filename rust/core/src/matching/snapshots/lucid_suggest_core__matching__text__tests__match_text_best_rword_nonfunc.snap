---
source: src/matching/text.rs
expression: "text_match(&rtext.to_ref(), &qtext.to_ref())"
---
(
    [
        WordMatch {
            offset: 1,
            slice: (
                4,
                9,
            ),
            subslice: (
                0,
                3,
            ),
            typos: 0.0,
            func: false,
            fin: false,
        },
    ],
    [
        WordMatch {
            offset: 0,
            slice: (
                0,
                3,
            ),
            subslice: (
                0,
                3,
            ),
            typos: 0.0,
            func: true,
            fin: false,
        },
    ],
)
